### PR TITLE
[mlir][spirv] Add gpu printf op lowering to spirv.CL.printf op

### PR DESCRIPTION
**Pull request by drprajap**
This change contains following:
	- adds lowering of printf op to spirv.CL.printf op in GPUToSPIRV pass.
	- Fixes Constant decoration parsing for spirv GlobalVariable.
	- minor modification to spirv.CL.printf op assembly format.

### CODE REVIEW

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 1]
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
...
+  unsigned stringNumber = 0;
+  mlir::SmallString<16> globalVarName;
+  mlir::spirv::GlobalVariableOp globalVar;
+
+  // formulate spirv global variable name
+  do {
+    globalVarName.clear();
+    (formatStringPrefix + llvm::Twine(stringNumber++))
+        .toStringRef(globalVarName);
+  } while (moduleOp.lookupSymbol(globalVarName));
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:640]: I don't understand what this is trying to do. Can we simplify this and/or add some comment with an example?

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 2]
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
...
+
+  auto i8Type = rewriter.getI8Type();
+  auto i32Type = rewriter.getI32Type();
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:643]:
```suggestion
  IntegerType i8Type = rewriter.getI8Type();
  IntegerType i32Type = rewriter.getI32Type();
```

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 3]
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
...
+
+  unsigned scNum = 0;
+  auto createSpecConstant = [&](unsigned value) {
+    auto attr = rewriter.getI8IntegerAttr(value);
+    mlir::SmallString<16> specCstName;
+    (llvm::Twine(globalVarName) + "_sc" + llvm::Twine(scNum++))
+        .toStringRef(specCstName);
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:650]: same as above

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 4]
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
...
+  do {
+    globalVarName.clear();
+    (formatStringPrefix + llvm::Twine(stringNumber++))
+        .toStringRef(globalVarName);
+  } while (moduleOp.lookupSymbol(globalVarName));
+
+  auto i8Type = rewriter.getI8Type();
+  auto i32Type = rewriter.getI32Type();
+
+  unsigned scNum = 0;
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:645]:
```suggestion
  unsigned specConstantNum = 0;
```

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 5]
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
...
+      rewriter.getBlock()->getParent()->getParentOfType<mlir::spirv::FuncOp>();
+
+  auto moduleOp = funcOp->getParentOfType<mlir::spirv::ModuleOp>();
+
+  const char formatStringPrefix[] = "printfMsg";
+  unsigned stringNumber = 0;
+  mlir::SmallString<16> globalVarName;
+  mlir::spirv::GlobalVariableOp globalVar;
+
+  // formulate spirv global variable name
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:635]:
```suggestion
  // Formulate SPIR-V global variable name.
```

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 6]
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
...
+    mlir::SmallString<16> specCstName;
+    (llvm::Twine(globalVarName) + "_sc" + llvm::Twine(scNum++))
+        .toStringRef(specCstName);
+
+    return rewriter.create<mlir::spirv::SpecConstantOp>(
+        loc, rewriter.getStringAttr(specCstName), attr);
+  };
+
+  // define GlobalVarOp with printf format string using SpecConstants
+  // and make composite of SpecConstants
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:657]:
```suggestion
  // Define a GlobalVarOp with printf format string using SpecConstants
  // and make composite of SpecConstants.
```

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 7]
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
...
+  {
+    mlir::Operation *parent =
+        mlir::SymbolTable::getNearestSymbolTable(gpuPrintfOp->getParentOp());
+
+    mlir::ConversionPatternRewriter::InsertionGuard guard(rewriter);
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:662]: nit: we don't need the mlir namespace prefix

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 8]
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
...
+
+    mlir::Block &entryBlock = *parent->getRegion(0).begin();
+    rewriter.setInsertionPointToStart(
+        &entryBlock); // insertion point at module level
+
+    // Create Constituents with SpecConstant to construct
+    // SpecConstantCompositeOp
+    llvm::SmallString<20> formatString(gpuPrintfOp.getFormat());
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:670]: Shouldn't this grab the format from the adaptor instead?

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 9]
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
...
+
+  // Get printf arguments
+  auto argsRange = adaptor.getArgs();
+  mlir::SmallVector<mlir::Value, 4> printfArgs;
+  printfArgs.reserve(argsRange.size() + 1);
+  printfArgs.append(argsRange.begin(), argsRange.end());
+
+  rewriter.create<mlir::spirv::CLPrintfOp>(loc, i32Type, fmtStr, printfArgs);
+
+  rewriter.eraseOp(gpuPrintfOp);
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:716]: Use `rewriter.replaceWithNewOp`.

File: mlir/test/Conversion/GPUToSPIRV/printf.mlir [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        // CHECK: [[FMTSTR_ADDR:%.*]] = spirv.mlir.addressof [[PRINTMSG]] : !spirv.ptr<!spirv.array<[[ARRAYSIZE]] x i8>, UniformConstant>
+        // CHECK-NEXT: [[FMTSTR_PTR1:%.*]] = spirv.Bitcast [[FMTSTR_ADDR]] : !spirv.ptr<!spirv.array<[[ARRAYSIZE]] x i8>, UniformConstant> to !spirv.ptr<i8, UniformConstant>
+        // CHECK-NEXT: {{%.*}} = spirv.CL.printf [[FMTSTR_PTR1]] : !spirv.ptr<i8, UniformConstant>({{%.*}}, {{%.*}}, {{%.*}} : i32, f32, i32) -> i32
+        gpu.printf "\nHello, world : %d %f \n Thread id: %d\n" %arg0, %arg1, %2: i32, f32, index
+        
+        // CHECK: spirv.Return
+        gpu.return
+    }
+  }
+}
```
Comment (kuhar) [inline mlir/test/Conversion/GPUToSPIRV/printf.mlir:71]: nit: missing newline

**Review by kuhar — CHANGES_REQUESTED** [review]
Thanks for the contribution, look like a useful debugging feature to have! I left some comments.

I find it hard to follow the `matchAndRewrite` code -- could we brake it down a bit and introduce some helper functions?

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 11]
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
...
+    globalVar->setAttr("Constant", rewriter.getUnitAttr());
+  }
+
+  // Get SSA value of Global variable
+  mlir::Value globalPtr =
+      rewriter.create<mlir::spirv::AddressOfOp>(loc, globalVar);
+  mlir::Value fmtStr = rewriter.create<mlir::spirv::BitcastOp>(
+      loc,
+      mlir::spirv::PointerType::get(i8Type,
+                                    mlir::spirv::StorageClass::UniformConstant),
```
Comment (MrSidims) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:705]: Should we instead check GV's storage class and use it instead of constant? Both SPIR-V and OpenCL have extensions relaxing printf string arg requirement, see https://github.com/KhronosGroup/SPIRV-Registry/pull/148

File: mlir/include/mlir/Dialect/SPIRV/IR/SPIRVCLOps.td [hunk 12]
```diff
@@ -875,7 +875,7 @@ def SPIRV_CLPrintfOp : SPIRV_CLOp<"printf", 184, []> {
     #### Example:
 
     ```mlir
-    %0 = spirv.CL.printf %0 %1 %2 : (!spirv.ptr<i8, UniformConstant>, (i32, i32)) -> i32
+    %0 = spirv.CL.printf %0 : !spirv.ptr<i8, UniformConstant>(%1, %2 : i32, i32) -> i32
```
[the remark this answers is not in the document]
Comment (drprajap) [reply mlir/include/mlir/Dialect/SPIRV/IR/SPIRVCLOps.td]: Thanks for suggestion. I changed it now.

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 13]
```diff
@@ -607,6 +616,108 @@ class GPUSubgroupReduceConversion final
...
+LogicalResult GPUPrintfConversion::matchAndRewrite(
+    gpu::PrintfOp gpuPrintfOp, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+
+  auto loc = gpuPrintfOp.getLoc();
+
+  auto funcOp =
+      rewriter.getBlock()->getParent()->getParentOfType<mlir::spirv::FuncOp>();
+
+  auto moduleOp = funcOp->getParentOfType<mlir::spirv::ModuleOp>();
```
[the remark this answers is not in the document]
Comment (drprajap) [reply mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp]: I added a check to see if module is valid then only proceed, let me know if it looks okay. This lowering is intended to be used with spirv.func op.

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp (hunk 9 above)
Comment (drprajap) [reply mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp]: I tried to use that met with a crash with below error
op->getNumResults() == newValues.size() && "incorrect # of replacement values"' failed
gpu.printf op doesn't take return type so I am not sure that creates an issue when using replaceOpWithNewOp.

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp (hunk 11 above)
Comment (drprajap) [reply mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp]: I am not sure I got your suggestion, do you mean we can get the storage class from already defined globalVar op? 

thanks for pointing out to the requirements docs, while implementing this, when I checked  OpenCL specs for printf , they mention fmt string needs to be in constant address space, hence I used UniformConstant here- https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_C.html#printf

**Comment by drprajap** [comment]
> Thanks for the contribution, look like a useful debugging feature to have! I left some comments.
> 
> I find it hard to follow the `matchAndRewrite` code -- could we brake it down a bit and introduce some helper functions?

Thank you @kuhar for your feedback and sorry for the delay in addressing them. 
I added more comments and tried to clarify the functionality, let me know if it makes it a bit more clear.

File: mlir/include/mlir/Dialect/SPIRV/IR/SPIRVCLOps.td [hunk 14]
```diff
@@ -875,7 +875,9 @@ def SPIRV_CLPrintfOp : SPIRV_CLOp<"printf", 184, []> {
     #### Example:
 
     ```mlir
-    %0 = spirv.CL.printf %0 %1 %2 : (!spirv.ptr<i8, UniformConstant>, (i32, i32)) -> i32
+    %0 = spirv.CL.printf %fmt %1, %2  : !spirv.ptr<i8, UniformConstant>, i32, i32 -> i32
+    
+   $format `,` ( $arguments^ )? attr-dict `:`  type($format) ( `,` type($arguments)^ )? `->` type($result)
```
Comment (kuhar) [inline mlir/include/mlir/Dialect/SPIRV/IR/SPIRVCLOps.td:880]: The assembly format shouldn't be in the examples section

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 15]
```diff
@@ -597,6 +606,120 @@ class GPUSubgroupReduceConversion final
...
+    gpu::PrintfOp gpuPrintfOp, OpAdaptor adaptor,
+    ConversionPatternRewriter &rewriter) const {
+
+  Location loc = gpuPrintfOp.getLoc();
+
+  auto moduleOp = gpuPrintfOp->getParentOfType<spirv::ModuleOp>();
+
+  if (!moduleOp) {
+    return success();
+  }
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:621]: `success` indicates that you matched, so we need to return `failure` here. Also some stylistic nits:
```suggestion
  auto moduleOp = gpuPrintfOp.getParentOfType<spirv::ModuleOp>();
  if (!moduleOp)
    return failure();
```

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 16]
```diff
@@ -597,6 +606,120 @@ class GPUSubgroupReduceConversion final
...
+
+  const char formatStringPrefix[] = "printfMsg";
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:623]:
```suggestion
  StringRef formatStringPrefix = "printfMsg";
```

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 17]
```diff
@@ -597,6 +606,120 @@ class GPUSubgroupReduceConversion final
...
+  // each global var needs to be created with a unique name.
+  // like printfMsg0, printfMsg1, ...
+  // Formulate unique global variable name after
+  // searching in the module for existing global variable names.
+  // This is to avoid name collision with existing global variables.
+  do {
+    globalVarName.clear();
+    (formatStringPrefix + llvm::Twine(stringNumber++))
+        .toStringRef(globalVarName);
+  } while (moduleOp.lookupSymbol(globalVarName));
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:639]: Could you make this into a helper function? This one is already a bit lengthy. Also, I'd use plain `std::string` here as the small size is less than what I'd expect the regular SSO length to be (20-ish characters). You can use `Twine::str()` to produce it.

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 18]
```diff
@@ -597,6 +606,120 @@ class GPUSubgroupReduceConversion final
...
+  unsigned specConstantNum = 0;
+  auto createSpecConstant = [&](unsigned value) {
+    auto attr = rewriter.getI8IntegerAttr(value);
+    SmallString<16> specCstName;
+    (llvm::Twine(globalVarName) + "_sc" + llvm::Twine(specConstantNum++))
+        .toStringRef(specCstName);
+
+    return rewriter.create<spirv::SpecConstantOp>(
+        loc, rewriter.getStringAttr(specCstName), attr);
+  };
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:658]: Similar here: could we outline this code to a helper function?

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 19]
```diff
@@ -597,6 +606,120 @@ class GPUSubgroupReduceConversion final
...
+  Value fmtStr = rewriter.create<spirv::BitcastOp>(
+      loc,
+      spirv::PointerType::get(i8Type, spirv::StorageClass::UniformConstant),
+      globalPtr);
+
+  // Get printf arguments
+  auto argsRange = adaptor.getArgs();
+  SmallVector<Value, 4> printfArgs;
+  printfArgs.reserve(argsRange.size() + 1);
+  printfArgs.append(argsRange.begin(), argsRange.end());
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:714]: Use `llvm::to_vector` if you need to

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 20]
```diff
@@ -597,6 +606,120 @@ class GPUSubgroupReduceConversion final
...
+
+  rewriter.create<spirv::CLPrintfOp>(loc, i32Type, fmtStr, printfArgs);
+
+  rewriter.eraseOp(gpuPrintfOp);
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:718]: Could you add a comment that we need this because the spirv op produces a result that we need to discard?

**Comment by kuhar** [comment]
@drprajap there are some test failures: https://buildkite.com/llvm-project/github-pull-requests/builds/101724#019202b6-5c05-43e1-aa01-48aba3a6a4de/6-202

**Comment by kuhar** [comment]
@drprajap could you click the 'Resolve' button on past comments that you believe have been addressed? This will make it easier for me to tell what are the remaining open discussion threads.

**Comment by drprajap** [comment]
> @drprajap there are some test failures: https://buildkite.com/llvm-project/github-pull-requests/builds/101724#019202b6-5c05-43e1-aa01-48aba3a6a4de/6-202

Yes, sorry I missed to update test cases with new printf format changes, updated now.

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp (hunk 17 above)
Comment (drprajap) [reply mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp]: Done, moved this to helper function that is used to construct globalVar and SpecConstant. It is using std::string and Twine::str() now.

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp (hunk 18 above)
Comment (drprajap) [reply mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp]: Done.

**Comment by drprajap** [comment]
> @drprajap could you click the 'Resolve' button on past comments that you believe have been addressed? This will make it easier for me to tell what are the remaining open discussion threads.

@kuhar I addressed all the feedback and resolved comments, please check it out and let me know if looks okay. Thanks so much for your valuable feedback.

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 21]
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
   }
 };
 
+/// Formulate unique variable/constant name after
+/// searching in the module for existing variable/constant names.
+/// This is to avoid name collision with existing variables.
+/// Example: printfMsg0, printfMsg1, printfMsg2, ...
+std::string constructVarName(spirv::ModuleOp moduleOp, llvm::StringRef prefix) {
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:613]:
```suggestion
static std::string makeVarName(spirv::ModuleOp moduleOp, llvm::StringRef prefix) {
```

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 22]
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
   }
 };
 
+/// Formulate unique variable/constant name after
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:609]:
```suggestion
/// Formulate a unique variable/constant name after
```

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 23]
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
...
+    ConversionPatternRewriter::InsertionGuard guard(rewriter);
+
+    Block &entryBlock = *parent->getRegion(0).begin();
+    rewriter.setInsertionPointToStart(
+        &entryBlock); // insertion point at module level
+
+    // Create Constituents with SpecConstant by scanning format string
+    // Each character of format string is stored as a spec constant
+    // and then these spec constants are used to create a
+    // SpecConstantCompositeOp
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:672]:
```suggestion
    // SpecConstantCompositeOp.
```

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 24]
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
...
+    llvm::SmallString<20> formatString(adaptor.getFormat());
+    formatString.push_back('\0'); // Null terminate for C
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:674]:
```suggestion
    formatString.push_back('\0'); // Null terminate for C.
```

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 25]
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
...
+    SmallVector<Attribute, 4> constituents;
+    for (auto c : formatString) {
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:676]:
```suggestion
    for (char c : formatString) {
```

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 26]
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
...
+      auto cSpecConstantOp = createSpecConstant(c);
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:677]: I wouldn't use `auto` here, the type is not obvious to me based on the RHS only.

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 27]
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
...
+
+  // Each character of printf format string is
+  // stored as a spec constant. We need to create
+  // unique name for this spec constant like
+  // @printfMsg0_sc0, @printfMsg0_sc1, ... by searching in the module
+  // for existing spec constant names.
+  auto createSpecConstant = [&](unsigned value) {
+    auto attr = rewriter.getI8IntegerAttr(value);
+    std::string specCstName =
+        constructVarName(moduleOp, (llvm::Twine(globalVarName) + "_sc").str());
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:654]: You can make `constructVarName` take `Twine` as the prefix string.

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 28]
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
...
+  // Get SSA value of Global variable and create pointer to i8 to point to
+  // the format string.
+  Value globalPtr = rewriter.create<spirv::AddressOfOp>(loc, globalVar);
+  Value fmtStr = rewriter.create<spirv::BitcastOp>(
+      loc,
+      spirv::PointerType::get(i8Type, spirv::StorageClass::UniformConstant),
+      globalPtr);
+
+  // Get printf arguments
+  SmallVector<Value, 4> printfArgs = llvm::to_vector<4>(adaptor.getArgs());
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:715]:
```suggestion
  auto printfArgs = llvm::to_vector_of<Value, 4>(adaptor.getArgs());
```

**Review by kuhar — COMMENTED** [review]
Overall looks good, just some remaining issues

**Comment by drprajap** [comment]
> Overall looks good, just some remaining issues

Thank you. Addressed remaining issues.

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 29]
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
...
+
+  // Each character of printf format string is
+  // stored as a spec constant. We need to create
+  // unique name for this spec constant like
+  // @printfMsg0_sc0, @printfMsg0_sc1, ... by searching in the module
+  // for existing spec constant names.
+  auto createSpecConstant = [&](unsigned value) {
+    auto attr = rewriter.getI8IntegerAttr(value);
+    std::string specCstName =
+        makeVarName(moduleOp, (llvm::Twine(globalVarName) + "_sc"));
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:654]:
```suggestion
        makeVarName(moduleOp, llvm::Twine(globalVarName) + "_sc");
```

File: mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp [hunk 30]
```diff
@@ -597,6 +606,124 @@ class GPUSubgroupReduceConversion final
...
+  }
+  // Get SSA value of Global variable and create pointer to i8 to point to
+  // the format string.
+  Value globalPtr = rewriter.create<spirv::AddressOfOp>(loc, globalVar);
+  Value fmtStr = rewriter.create<spirv::BitcastOp>(
+      loc,
+      spirv::PointerType::get(i8Type, spirv::StorageClass::UniformConstant),
+      globalPtr);
+
+  // Get printf arguments
```
Comment (kuhar) [inline mlir/lib/Conversion/GPUToSPIRV/GPUToSPIRV.cpp:714]:
```suggestion
  // Get printf arguments.
```

**Review by kuhar — APPROVED** [review]
LGTM

File: mlir/test/Conversion/GPUToSPIRV/printf.mlir [hunk 31]
```diff
@@ -0,0 +1,71 @@
+// RUN: mlir-opt -allow-unregistered-dialect -split-input-file  -convert-gpu-to-spirv -verify-diagnostics %s -o - | FileCheck %s
```
Comment (kuhar) [inline mlir/test/Conversion/GPUToSPIRV/printf.mlir:1]: Do we need the `-o o`?

Comment (drprajap) [inline mlir/test/Conversion/GPUToSPIRV/printf.mlir:1]:
```suggestion
// RUN: mlir-opt -allow-unregistered-dialect -split-input-file  -convert-gpu-to-spirv -verify-diagnostics %s | FileCheck %s
```

Comment (drprajap) [reply mlir/test/Conversion/GPUToSPIRV/printf.mlir:1]: Not really, was just trying to follow test command lines from other tests, I can remove it.

**Comment by drprajap** [comment]
> LGTM

Thanks, I do not have write permission, can you please help merge it if possible? else I can ask other team members.

**Comment by kuhar** [comment]
> > LGTM
> 
> Thanks, I do not have write permission, can you please help merge it if possible? else I can ask other team members.

Sure, I will merge this on Monday. Please ping me if I forget to do so.

**Comment by drprajap** [comment]
> > > LGTM
> > 
> > 
> > Thanks, I do not have write permission, can you please help merge it if possible? else I can ask other team members.
> 
> Sure, I will merge this on Monday. Please ping me if I forget to do so.

Gentle ping to merge this. Thank you :)
